### PR TITLE
[Android] Fix RefreshView loading indicator position

### LINKED ISSUE #17592: [Android] RefreshView disappears when switching tabs in a TabbedPage
### Description

When a RefreshView is hosted inside a TabbedPage, the RefreshView's progress circle disappears when switching between tabs.

Also, as a bonus bug: why is the RefreshView's progress circle half occluded by the tabs?

image

### Steps to Reproduce

1. Get a RefreshView spinning in a tab.
2. Switch to the other tab. Go back.
3. The RefreshView is still refreshing behind the scenes, but its progress circle has vanished

### Link to public reproduction project repository

https://github.com/mfeingol/repros/tree/master/RefreshViewInTabbedPageRepro

### Version with bug

8.0.0-rc.1.9171

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

13

### Did you find any workaround?

No.

### Relevant log output

_No response_

### LINKED ISSUE #17610: [regression/8.0.0-rc.1.9171] RefreshView indicator hidden behind Navigation bar
### Description

When using an RefreshView at .net 8 the indicator at refreshing is always hidden behind navigation bar, before at .net 7 I can see it completly.

.net 8
behind

.net 7
behind7

```  
<ContentPage.Content>
     <RefreshView  >
             <CollectionView 
                  ItemsSource="{Binding Employees}" 
                  ItemTemplate="{StaticResource EmployeeTemplate}" 
                  ItemsLayout="VerticalGrid, 2" />
         </RefreshView>
        

     
 </ContentPage.Content>
 ```

### Steps to Reproduce

_No response_

### Link to public reproduction project repository

_No response_

### Version with bug

[8.0.0-rc.1.9171](https://github.com/dotnet/maui/releases/tag/8.0.0-rc.1.9171)

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

[8.0.0-preview.7.8842](https://github.com/dotnet/maui/releases/tag/8.0.0-preview.7.8842)

### Affected platforms

Android

### Affected platform versions

Android 13

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #18987: RefreshView indicator is begind Content Page Title
### Description

RefreshView indicator is behind Content Page Title

### Steps to Reproduce

Create .NET 8 MAUI App
Put RefreshView inside a Tabbed page
Title hover the RefreshView indicator

Screenshot_20231123-160134 1

### Link to public reproduction project repository

_No response_

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

6.0

### Affected platforms

Android

### Affected platform versions

Android 13

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

Fix RefreshView loading indicador position on Android.

Before the changes
issue-17610-before
After the changes
issue-17610-after

.NET MAUI Gallery RefreshView sample behavior with the changes
refreshview

### Issues Fixed

Fixes #17610
Fixes #18987

### CODE REVIEW

File: src/Core/src/Handlers/RefreshView/RefreshViewHandler.Android.cs
Comment (jsuarezruiz) [inline src/Core/src/Handlers/RefreshView/RefreshViewHandler.Android.cs:17]: The refresh indicator resting position is always positioned near the top of the refreshing content. With the changes, we always set the position 50 px based on whether or not there is a toolbar or actionbar present (ProgressViewEndOffset value).

**Review by RelxOff — APPROVED** [review]
👍

File: src/Core/src/Handlers/RefreshView/RefreshViewHandler.Android.cs
Comment (dustin-wojciechowski) [reply src/Core/src/Handlers/RefreshView/RefreshViewHandler.Android.cs:17]: Hmm, a solution that incorporated changed the px count according to device dpi would be a more rigid solution to this, I think. But the 50 was just enough to clear the issue of the spinner getting stuck. This was a strange issue because I'm certain it's a fault on the Android side.

**Review by dustin-wojciechowski — APPROVED** [review]
An explanatory comment could benefit if we need to revisit at a later date (or even a variable for user to modify themselves for that offset), but solution should handle this problem.

**Review by PureWeen — CHANGES_REQUESTED** [review]
I'm guessing the 50 comes from the action bar height which is 48px. We have a `ContextExtentions` property that's GetActionBarHeight()

Though I do wonder about this fix somewhat. 

If you inspect the layout using Android Studio does the refreshview extend up into the ActionBar? Is that why the visualization is under the actionbar? 

Do we need to fix the REfreshView so that it's bounds don't go underneathe the action bar?

**Comment by jsuarezruiz** [comment]
> I'm guessing the 50 comes from the action bar height which is 48px. We have a `ContextExtentions` property that's GetActionBarHeight()
> 
> Though I do wonder about this fix somewhat.
> 
> If you inspect the layout using Android Studio does the refreshview extend up into the ActionBar? Is that why the visualization is under the actionbar?
> 
> Do we need to fix the REfreshView so that it's bounds don't go underneathe the action bar?

The `ActionBarHeight `is bigger than the default `ProgressViewEndOffset`.

File: src/Core/src/Handlers/RefreshView/RefreshViewHandler.Android.cs
Comment (PureWeen) [reply src/Core/src/Handlers/RefreshView/RefreshViewHandler.Android.cs:17]: https://issuetracker.google.com/issues/110463864

Comment (PureWeen) [reply src/Core/src/Handlers/RefreshView/RefreshViewHandler.Android.cs:17]: It looks like there are a few things going on here

1) it looks like there's a bug with the SwipeRefreshLayout.  It's kind of odd because they fixed it two years ago, but it's not fixed in the latest androidx packages. If you look at the main branch they've added an else into the nestedscrollview code https://android-review.googlesource.com/c/platform/frameworks/support/+/1734993 which triggers a reset of the indicator so it hides when you scroll up. I tried a few paths to mimic what they are doing but I'm not sure a reliable way to do this.

2) I'm not sure why exactly, but, for whatever reason our RefreshView isn't clipping to bounds. If you use the SwipeRefreshLayout in a platform app you'll see the indicator goes outside the bounds of the refreshview so it hides itself out of view. My guess is that this is related to our layouts vs using androids in the box layouts. 

3) Calling any of the SetProgress* APIs works around the [original](https://github.com/dotnet/maui/pull/17080) issue because calling any of those APIs sets the `mUsingCustomStart` private variable on the SwipeRefreshLayout which causes this path, so the visibility of the spinner will get set to `Gone` once you've scrolled all the way back to zero. 

image

The size of the ToolBar is kind of irrelevant here. Like if you just call this
`SetProgressViewOffset(false, ProgressViewStartOffset, ProgressViewEndOffset);`

It works "fine"

if you call this
`SetProgressViewOffset(false, 0, ProgressViewEndOffset);`

It also works fine and then the refresh circle doesn't go above the `RefreshView`.

I'm not quite sure where y'all got 50 from, and what that's trying to accomplish? 

If we want to somewhat replicate the values inside the `SwipeRefreshLayout` then I think we'd call it like this

`SetProgressViewOffset(true, ProgressViewStartOffset, ProgressViewEndOffset - Math.Abs(ProgressViewStartOffset));`

Which I just pulled from here

image

That causes the indicator when refreshing to land in the same spot. I like setting the scale to true. I think the non scaling version is odd, especially if there's a toolbar.


Also, I would call `SetProgressViewOffset` inside the ctor of `SwipeRefreshLayout`, that way if developers want to call `SetProgressViewOffset` themselves they don't have to figure out when we call it.

Comment (PureWeen) [reply src/Core/src/Handlers/RefreshView/RefreshViewHandler.Android.cs:17]: Once we merge this if we could add a manual test that'd be cool!

**Review by PureWeen — CHANGES_REQUESTED** [review]
https://github.com/dotnet/maui/pull/17647#discussion_r1433358418

I think we should put this
`SetProgressViewOffset(true, ProgressViewStartOffset, ProgressViewEndOffset - Math.Abs(ProgressViewStartOffset));`

into the ctor of our MauiSwipeRefreshLayout

That seems to yield a behavior closest to the default.

**Comment by jsuarezruiz** [comment]
Close in favor of https://github.com/dotnet/maui/pull/19767

**Pull request closed without merging**
